File: android/app/src/main/kotlin/space/go2china/visepanda/data/model/CanvasPatchApplier.kt
package space.go2china.visepanda.data.model

object CanvasPatchApplier {

    fun apply(current: TripState, patch: CanvasPatch): TripState {
        val alertMap = LinkedHashMap<String, ButlerAlert>()
        current.alerts.forEach { alert ->
            alertMap[alert.dedupeKey()] = alert
        }
        patch.butlerAlerts.orEmpty().forEach { alert ->
            alertMap[alert.dedupeKey()] = alert
        }

        return current.copy(
            summary = current.summary.merge(patch.tripSummary),
            days = patch.days ?: current.days,
            alerts = alertMap.values.toList(),
            lastUpdatedReason = patch.reason,
        )
    }

    private fun TripSummary.merge(patch: TripSummaryPatch?): TripSummary {
        if (patch == null) return this
        return copy(
            title = patch.title ?: title,
            durationDays = patch.durationDays ?: durationDays,
            pace = patch.pace ?: pace,
            travelerStyle = patch.travelerStyle ?: travelerStyle,
            destinations = patch.destinations ?: destinations,
            confidence = patch.confidence ?: confidence,
        )
    }

    private fun ButlerAlert.dedupeKey(): String = "${type.name}:$title"
}

File: ios/VisePandaIOS/Data/CanvasPatchApplier.swift
import Foundation

enum CanvasPatchApplier {
    static func apply(current: TripState, patch: CanvasPatch) -> TripState {
        var summary = current.summary
        if let tripSummary = patch.tripSummary {
            summary.title = tripSummary.title ?? summary.title
            summary.durationDays = tripSummary.durationDays ?? summary.durationDays
            summary.pace = tripSummary.pace ?? summary.pace
            summary.travelerStyle = tripSummary.travelerStyle ?? summary.travelerStyle
            summary.destinations = tripSummary.destinations ?? summary.destinations
            summary.confidence = tripSummary.confidence ?? summary.confidence
        }

        var alertMap = Dictionary(uniqueKeysWithValues: current.alerts.map { ($0.id, $0) })
        for alert in patch.butlerAlerts ?? [] {
            alertMap[alert.id] = alert
        }

        return TripState(
            summary: summary,
            days: patch.days ?? current.days,
            alerts: Array(alertMap.values).sorted { $0.title < $1.title },
            lastUpdatedReason: patch.reason
        )
    }
}
